java/flight/src/main/java/org/apache/arrow/flight/FlightServer.java

```diff
@@ -227,13 +236,19 @@ public FlightServer build() {
       }
 
       // Share one executor between the gRPC service, DoPut, and Handshake
+      final List<AutoCloseable> resources = new ArrayList<>();
       final ExecutorService exec = executor != null ? executor : Executors.newCachedThreadPool();
+      if (executor == null) {
+        // If we create an executor, make sure to shut it down alongside the server.
+        resources.add(exec::shutdownNow);
```

Comment:
why shutdownNow and not shutdown?

Reply:
I was worried about it blocking but on reflection, if we make it to closing the executor, the gRPC server should be shut down and everything should be terminated. I've changed it and updated the comment to explain.